cpu/msp430_common/cpu.c

```diff
@@ -14,23 +14,46 @@
 #include "thread.h"
 
 /*
- * we must prevent the compiler to generate a prologue or an epilogue
- * for thread_yield_higher(), since we rely on the RETI instruction at the end
- * of its execution, in the inlined __restore_context() sub-function
+ * This function can both be called from ISR and from thread context.
+ *
+ * In both cases, the caller will use "CALL", which pushes the return
+ * address on the stack before executing the function's first instruction.
+ *
+ * If called within ISR, just set sched_context_switch_request and
+ * directly return to the call site. A regular "ret" does this.
+ *
+ * If called from stack context, it needs to prepare the stack so it looks exactly
+ * as if the thread had been interrupted by an ISR, which requires the SR to be on
+ * stack right after the return address. So we do this manually.
+ * __save_context() will then save the remaining registers, then store the stack
+ * pointer in the thread's thread_t.
+ *
+ * At this point, the thread context is properly saved. sched_run (possibly) changes
+ * the currently active thread, which __restore_context() then restores, resuming
+ * execution at the call site using reti.
+ *
+ * The naked attribute is needed to ensure this works.
  */
 __attribute__((naked)) void thread_yield_higher(void)
 {
-    __asm__("push r2"); /* save SR */
-    irq_disable();
+    if (irq_is_in()) {
+        sched_context_switch_request = 1;
+        __asm__("ret");
+    }
+    else {
+        __asm__("push r2"); /* save SR */
+        __asm__("dint");
+        __asm__("nop");
```

Comment:
Is it normal that interrupts are disabled and not restored?

Reply:
explained offline

Reply:
> explained offline

Now I'm curious

Reply:
```The "reti" restores the status register from the saved context. "Push r2" saves it there```

Reply:
they are implicitly enabled by the inlined "reti" in __restore_context() by restoring the SR register.